Replace graphql-docs npm package with gem

alecslupu opened the pull request:
#### :tophat: What? Why?
Some NPM packages used by the project are very old. This could affect the project maintenaince and security, as it's not possible to update packages with recent versions. This PR tries to improve that.

These are the main issues found:
- The `graphql-docs` npm package is very old and it's not maintained. I've tried to update it but it requires too much work to do it. I've found a gem also called `graphql-docs` that generates a very similar output but statically. I've created a rake task that uses it and generates those static files in the `app/views/static/` folder. Then, I've embeded those files into the docs view, to get a result very similar to the existing solution. The task can be run manually, but it will be run after the creation of the app and after each decidim upgrade. The only "problem" is that this static files should be added to the app repository, and that's a bit ugly.

Most of the work has been done by @leio10 in #8335

#### :pushpin: Related Issues
*Link your PR to an issue*
- Related to #8335

- Related to https://github.com/gjtorikian/graphql-docs/pull/89
- Related to https://github.com/gjtorikian/graphql-docs/pull/88


### :camera: Screenshots
*Please add screenshots of the changes you're proposing*
![image](https://user-images.githubusercontent.com/105683/146736484-b8f50a2f-8041-495b-8b4f-584dfdcac017.png)
![image](https://user-images.githubusercontent.com/105683/146736525-2a9f76a4-b9b6-4a5a-a2e5-1bfbdbc9b90e.png)


:hearts: Thank you!

alecslupu commented on decidim-generators/spec/generators_spec.rb:
```diff
@@ -109,7 +109,7 @@ module Decidim
 
       context "with --branch flag" do
-        let(:default_branch) { "develop" }
+        let(:default_branch) { "ale/graphql-api" }
```
To be changed once it's approved (I'm leaving it, so we don't miss it)

alecslupu commented on decidim-generators/lib/decidim/generators/app_generator.rb:
```diff
@@ -262,7 +262,7 @@ def target_gemfile
                  expanded_path
                elsif branch.present?
-                 "https://raw.githubusercontent.com/decidim/decidim/#{branch}/decidim-generators"
+                 "https://raw.githubusercontent.com/tremend-cofe/decidim/#{branch}/decidim-generators"
```
To be changed once it's approved (I'm leaving it, so we don't miss it)

alecslupu commented on decidim-generators/lib/decidim/generators/app_generator.rb:
```diff
@@ -237,7 +237,7 @@ def gem_modifier
         return if options[:path]
 
-        @branch ||= options[:edge] ? "develop" : options[:branch].presence
+        @branch ||= options[:edge] ? "ale/graphql-api" : options[:branch].presence
```
To be changed once it's approved (I'm leaving it, so we don't miss it)

alecslupu commented on decidim-generators/lib/decidim/generators/app_generator.rb:
```diff
@@ -228,7 +228,7 @@ def gem_modifier
                             "path: \"#{options[:path]}\""
                           elsif branch.present?
-                            "git: \"https://github.com/decidim/decidim.git\", branch: \"#{branch}\""
+                            "git: \"https://github.com/tremend-cofe/decidim.git\", branch: \"#{branch}\""
```
To be changed once it's approved (I'm leaving it, so we don't miss it)

andreslucena commented:
> Most of the work has been done by @leio10 in #8335 

On this kind of PRs, where we supersede work done by others and using that as the base, I prefer to have the original commits cherry-picked (or even directly using the original branch as a starter). This is important for respecting the authorship of the original developer. 

Is there any other step to be done for seeing this? Locally, after creating the development_app, I see this:

![image](https://user-images.githubusercontent.com/717367/148032696-d946e54b-fa7e-4e5c-9c2b-a15eb685f4ec.png)

I don't see any js nor rails exception

alecslupu commented:
> On this kind of PRs, where we supersede work done by others and using that as the base, I prefer to have the original commits cherry-picked (or even directly using the original branch as a starter). This is important for respecting the authorship of the original developer.
@andreslucena, That particular PR has 22 commits, and i found it hard to cherry-pick the required commits. 

On the other hand, in order to test this, you will need to run the following commands: 
```
rake decidim:upgrade
rake decidim_api:generate_docs
```

additionally, you may want to remove the following: 
```
rm -rf public/decidim-packs/
rm -rf  app/views/static/
```

alecslupu commented:
Current coverage: 
![image](https://user-images.githubusercontent.com/105683/148062994-c4de13c9-3645-4f84-9629-be8c2da6f1e5.png)

andreslucena commented:
> On the other hand, in order to test this, you will need to run the following commands (after you refetch)

Ok, I tried with a new `development_app`, as far as I see it should run those tasks automatically... I've run them manually (and also with another new `development_app`) and now the only style that I see missing is the one for the sidebar (at the bottom of the page, `#sidebar`)

![image](https://user-images.githubusercontent.com/717367/148897845-e2a720ed-0743-410e-a162-49de61878932.png)

alecslupu commented:
While I rebased the work to fix the conflicts i have tried to replicate the steps. 

After the change and rebase, i have accessed the application and i have seen something like the pics you have attached on the PR. After I have done a CTRL+F5 the page looked normally. 
I have added a before action to make sure nothing is cached at the browser level.

andreslucena commented:
@alecslupu - now it's working great, thanks!!

andreslucena commented on decidim-api/lib/decidim/api.rb:
```diff
@@ -7,6 +7,23 @@ module Decidim
+      50
+    end
+
+    # defines the schema max_per_page to configure GraphQL query complexity
```
max_per_page?

alecslupu replied:
copy/paste programming :spaghetti:

alecslupu replied:
# defines the schema **max_complexity** to configure GraphQL query complexity

andreslucena commented on decidim-api/lib/decidim/api.rb:
```diff
@@ -7,6 +7,23 @@ module Decidim
+      5000
+    end
+
+    # defines the schema max_per_page to configure GraphQL query max_depth
```
max_per_page?

alecslupu replied:
# defines the schema **max_per_page** to configure GraphQL query max_depth

andreslucena commented on decidim-api/spec/tasks/decidim_api_generate_docs_spec.rb:
```diff
@@ -0,0 +1,16 @@
+
+describe "rake decidim_api:generate_docs", type: :task do
+  it "creates the static docs files" do
+    static = Rails.root.join("app/views/static/api/docs")
```
Should we add something in the CHANGELOG? It could be weird that for a Decidim upgrade, apart from the typical file changes that we usually do (`Gemfile, Gemfile.lock, db/migrate/*, db/schema.rb`) that we also add these files without saying anything to implementers.

alecslupu replied:
![image](https://user-images.githubusercontent.com/105683/149167477-1bf8f823-003c-4321-b96d-1f295413ca20.png)

andreslucena commented:
Ready to be merged for me, great job 👏🏽 👏🏽 

Please change the branch to `develop` so I can merge it.

alecslupu commented:
> Ready to be merged for me, great job 👏🏽 👏🏽
> 
> Please change the branch to `develop` so I can merge it.

@andreslucena , i have added f721f0dcb404f2dbb6a512046534b93fe0cbfc6c that makes the repository switch back to Decidim Core.

andreslucena commented on CHANGELOG.md:
````diff
@@ -4,6 +4,24 @@
+
+The static documentation will be rendered into : ```app/views/static/api/docs``` which is being refreshed automatically when you will run ```rake decidim:upgrade```.
+
+You can automatically regenerate the docs by running: ```rake decidim_api:generate_docs```
````
```suggestion
You can manually regenerate the docs by running: ```rake decidim_api:generate_docs```
```

andreslucena reviewed (changes requested):
A minor change (the last one, I promise 😄)

andreslucena commented:
Seems like the CI error in Generators is relevant @alecslupu:

>        Don't know how to build task 'decidim_api:generate_docs' (See the list of available tasks with `rails --tasks`) 

Can you check it?

alecslupu commented:
> Seems like the CI error in Generators is relevant @alecslupu:
> 
> > ```
> >    Don't know how to build task 'decidim_api:generate_docs' (See the list of available tasks with `rails --tasks`) 
> > ```
> 
> Can you check it?

The generator errors is caused by the fact that specs cannot find the task in the develop branch. 

Check f721f0d that makes the repository switch back to Decidim Core, and that is actually breaks the generators specs.

andreslucena commented:
> The generator errors is caused by the fact that specs cannot find the task in the develop branch. 

That makes perfect sense 😅
